src/test/kotlin/cacophony/regex/RegexParserTest.kt

```diff
@@ -0,0 +1,253 @@
+        val s2 = ConcatenationRegex(StarRegex(AtomicRegex('x')), ConcatenationRegex(AtomicRegex('a'), StarRegex(AtomicRegex('b'))))
+        val expected: AlgebraicRegex = UnionRegex(s0, s1, s2)
+        assertEqualAlgebraicRegex(result, expected)
+    }
+
+    @Test
+    fun `useless parenthesis`() {
+        val result = parseRegex("""(((a))((b))(c)|(x)(((((a))))|(((b))))|(((x)))*((a)(((b)))*))""")
+        val s0 = ConcatenationRegex(*"""abc""".map { AtomicRegex(it) }.toTypedArray())
+        val s1 = ConcatenationRegex(AtomicRegex('x'), UnionRegex(AtomicRegex('a'), AtomicRegex('b')))
+        val s2 = ConcatenationRegex(StarRegex(AtomicRegex('x')), ConcatenationRegex(AtomicRegex('a'), StarRegex(AtomicRegex('b'))))
+        val expected: AlgebraicRegex = UnionRegex(s0, s1, s2)
+        assertEqualAlgebraicRegex(result, expected)
+    }
+
+    @Test
+    fun `actual keywords`() {
+        val keywords = arrayOf("""let""", """while""", """do""", """if""", """then""", """else""", """true""", """false""")
+        val result = parseRegex(keywords.joinToString("""|"""))
+        val expected: AlgebraicRegex =
+            UnionRegex(
+                *keywords.map {
+                    ConcatenationRegex(
+                        *it.map {
+                                c ->
+                            AtomicRegex(c)
+                        }.toTypedArray(),
+                    )
+                }.toTypedArray(),
+            )
+        assertEqualAlgebraicRegex(result, expected)
+    }
+
+    @Test
+    fun `non special operators`() {
+        val operators = """[]&=<>+-!#%"""
+        val result = parseRegex(operators.toCharArray().joinToString("""|"""))
+        val expected: AlgebraicRegex = UnionRegex(*operators.map { AtomicRegex(it) }.toTypedArray())
+        assertEqualAlgebraicRegex(result, expected)
+    }
+
+    @Test
+    fun `special operators`() {
+        val operators = """|*\()"""
+        val result = parseRegex(operators.toCharArray().joinToString("""|\""", """\"""))
+        val expected: AlgebraicRegex = UnionRegex(*operators.map { AtomicRegex(it) }.toTypedArray())
+        assertEqualAlgebraicRegex(result, expected)
+    }
+
+    @Test
+    fun `escaped characters`() {
+        val characters = """nrt"""
+        val result = parseRegex(characters.toCharArray().joinToString("""|\""", """\"""))
+        val expected: AlgebraicRegex = UnionRegex(AtomicRegex('\n'), AtomicRegex('\r'), AtomicRegex('\t'))
+        assertEqualAlgebraicRegex(result, expected)
+    }
+
+    @Test
+    fun `do not recognize empty word`() {
+        assertThrows<RegexSyntaxErrorException> { parseRegex("abc|()|def") }
+    }
+
+    @Test
+    fun `do not recognize empty word here either`() {
+        assertThrows<RegexSyntaxErrorException> { parseRegex("abc|x()z|def") }
+    }
+
+    @Test
+    fun `mismatched parenthesis 1`() {
+        assertThrows<RegexSyntaxErrorException> { parseRegex("abc|x(z|d))ef") }
```

Comment:
You shouldn't name tests like "doing something 1/2/3...", you can add few asserts to the same test as long as it's similar case

Reply:
Renamed